src/agent/coverage/src/block/linux.rs

```diff
@@ -0,0 +1,499 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::ffi::OsStr;
+use std::path::{Path, PathBuf};
+
+use anyhow::Result;
+use iced_x86::{Decoder, DecoderOptions, Instruction};
+use object::endian::LittleEndian as LE;
+use object::{read::elf, Object, ObjectSection, ObjectSegment, ObjectSymbol};
+use pete::{Command, Ptracer, Restart, Signal, Stop, Tracee};
+use procfs::process::{MMapPath, MemoryMap, Process};
+
+use crate::block::ModuleCov;
+
+pub fn record(cmd: Command) -> Result<CommandBlockCov> {
+    let mut recorder = Recorder::default();
+    recorder.record(cmd)?;
+    Ok(recorder.coverage)
+}
+
+#[derive(Default, Debug)]
+pub struct Recorder {
+    breakpoints: Breakpoints,
+    coverage: CommandBlockCov,
+    images: Option<Images>,
+}
+
+impl Recorder {
+    pub fn record(&mut self, cmd: Command) -> Result<()> {
+        use pete::ptracer::Options;
+
+        let mut ptracer = Ptracer::new();
+
+        // Attach stop.
+        let mut tracee = ptracer.spawn(cmd)?;
+
+        ptracer.restart(tracee, Restart::Continue)?;
+
+        // Continue until `exec()`.
+        while let Some(t) = ptracer.wait()? {
+            if let Stop::Exec(..) = t.stop {
+                tracee = t;
+                break;
+            }
+
+            ptracer.restart(tracee, Restart::Continue)?;
+        }
+
+        // Do not follow forks.
+        //
+        // After this, we assume that any new tracee is a thread in the same
+        // group as the root tracee.
+        let mut options = Options::all();
+        options.remove(Options::PTRACE_O_TRACEFORK);
+        options.remove(Options::PTRACE_O_TRACEVFORK);
+        options.remove(Options::PTRACE_O_TRACEEXEC);
+        tracee.set_options(options)?;
+
+        self.images = Some(Images::new(tracee.pid.as_raw()));
+        self.update_images(&mut tracee)?;
+
+        ptracer.restart(tracee, Restart::Syscall)?;
+
+        while let Some(mut tracee) = ptracer.wait()? {
+            match tracee.stop {
+                Stop::SyscallEnterStop(..) => log::trace!("syscall-enter: {:?}", tracee.stop),
+                Stop::SyscallExitStop(..) => {
+                    self.update_images(&mut tracee)?;
+                }
+                Stop::SignalDeliveryStop(_pid, Signal::SIGTRAP) => {
+                    self.on_breakpoint(&mut tracee)?;
+                }
+                Stop::Clone(pid, tid) => {
+                    log::info!("new thread: {} -> {}", pid, tid);
+                }
+                _ => {
+                    log::debug!("stop: {:?}", tracee.stop);
+                }
+            }
+
+            if let Err(err) = ptracer.restart(tracee, Restart::Syscall) {
+                log::error!("unable to restart tracee: {}", err);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn update_images(&mut self, tracee: &mut Tracee) -> Result<()> {
+        let events = self.images.as_mut().unwrap().update()?;
```

Comment:
These unwraps are assume the method is private and preconditions are met. Convert to `Result` values and use `?` instead.